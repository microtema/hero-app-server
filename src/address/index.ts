import {Address} from './Address';
import AddressRepository from './Address.repository';
import AddressService from './Address.service';

export {Address, AddressService, AddressRepository};
