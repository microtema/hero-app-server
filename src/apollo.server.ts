import {ApolloServer as ApolloServerExpress, gql} from 'apollo-server-express';
import {Inject, Singleton} from 'typescript-ioc';
import AuthorService from './author/AuthorService';
import BookService from './book/BookService';

@Singleton
export default class ApolloServer {

    private static instance: ApolloServer;
    public readonly server: ApolloServerExpress;

    public readonly resolvers = {
        Query: {
            author(parent, {id}, {req}, info) {

                return ApolloServer.instance.authorService.getAuthor(id);
            },
            authors(parent, {name}, {req}, info) {

                return ApolloServer.instance.authorService.getAuthors({name});
            },
            book(parent, {id}, {req}, info) {

                return ApolloServer.instance.bookService.getBook(id);
            },
            books(parent, {title}, {req}, info) {

                return ApolloServer.instance.bookService.getBooks({title});
            },
        },
    };

    public readonly typeDefs = gql`
        type Query {
            author(id: ID): Author
            authors(name: String): [Author]
            book (id: ID): Book
            books (title: String): [Book]
        }

        type Book {
            id: ID
            title: String
            author: Author
        }

        type Author {
            id: ID
            name: String
            books: [Book]
        }
    `;

    public readonly config = {
        context: ({req}) => ({req}), // pass in the request to allow for authorization
        resolvers: this.resolvers,
        typeDefs: this.typeDefs,
    };

    constructor(@Inject private authorService: AuthorService, @Inject private bookService: BookService) {

        this.server = new ApolloServerExpress(this.config);

        // NOTE This is important on resolvers, since the scope of Query functions is not this.
        ApolloServer.instance = this;
    }
}
