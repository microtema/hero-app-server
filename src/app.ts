import express from 'express';

const app = express();

app.get('/', (req, res) => {

    res.send('The sedulous hyena ate the antelope!');
});

export default app;
