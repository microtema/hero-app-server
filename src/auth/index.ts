import AuthService from './Auth.service';

export {AuthService};
