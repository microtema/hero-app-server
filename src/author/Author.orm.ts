import {EntityStatic} from '../repository/Entity';

const AuthorModel = (sequelize, DataTypes): EntityStatic<number> => {

    const model = sequelize.define('author', {
        name: {
            allowNull: false,
            type: DataTypes.STRING,
        },
    });

    model.associate = (models) => {
        model.hasMany(models.BookModel, {onDelete: 'CASCADE'});
        model.hasMany(models.AddressModel, {onDelete: 'CASCADE'});
    };

    return model as EntityStatic<number>;
};

export default AuthorModel;
