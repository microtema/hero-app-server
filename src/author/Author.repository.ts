import {Op} from 'sequelize';
import {Singleton} from 'typescript-ioc';
import CrudRepository from '../repository/CrudRepository';
import models from '../repository/models';
import {Author} from './Author';

@Singleton
export default class AuthorRepository extends CrudRepository<Author, number> {

    /**
     * Provide AuthorModel
     */
    public model() {

        return models.AuthorModel;
    }

    /**
     * Find all Authors filtered by name
     *
     * @param name may be undefined or null
     */
    public findAll(name?: string) {

        const options = {
            include: [models.BookModel, models.AddressModel],
            where: {
                name: {
                    [Op.iLike]: '%' + (name || ''),
                },
            },
        };

        return super.findAll(options);
    }

    /**
     * Find Author by name
     *
     * @param id may not be null
     */
    public findByPk(id: number) {

        return super.findByPk(id, {include: [models.BookModel, models.AddressModel]});
    }
}
