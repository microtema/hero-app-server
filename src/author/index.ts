import AuthorBuilder from './AuthorBuilder';
import AuthorService from './AuthorService';

export {AuthorBuilder, AuthorService};
