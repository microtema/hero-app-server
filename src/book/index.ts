import BookBuilder from './BookBuilder';
import BookService from './BookService';

export {BookBuilder, BookService};
