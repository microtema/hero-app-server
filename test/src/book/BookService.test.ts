import {AuthorService} from '../../../src/author';
import {BookService} from '../../../src/book';

describe('Test AuthorService', () => {

    const authorService = {
        getAuthors: jest.fn((size: number) => [{
            books: [{
                id: '1000',
                title: 'AWS',
            }],
        }]) as any,
    } as AuthorService;

    const sut = new BookService(authorService);

    it('It should be defined', () => {

        expect(sut).toBeDefined();
    });

    it('getBook by id should match', () => {

        expect(sut.getBook('1000')).toMatchSnapshot();

        expect(authorService.getAuthors).toBeCalledWith({name: ''});
    });

    it('getBook by id should not match', () => {

        expect(sut.getBook('0000')).toMatchSnapshot();
    });

    it('getBooks by title should match', () => {

        expect(sut.getBooks({title: 'AWS'})).toMatchSnapshot();

        expect(authorService.getAuthors).toBeCalledWith({name: ''});
    });

    it('getBooks by title should  not match', () => {

        expect(sut.getBooks({title: '0000'})).toMatchSnapshot();
    });
});
