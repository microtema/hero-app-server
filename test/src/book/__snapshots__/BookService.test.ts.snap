// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test AuthorService getBook by id should match 1`] = `
Object {
  "id": "1000",
  "title": "AWS",
}
`;

exports[`Test AuthorService getBook by id should not match 1`] = `undefined`;

exports[`Test AuthorService getBooks by title should  not match 1`] = `Array []`;

exports[`Test AuthorService getBooks by title should match 1`] = `
Array [
  Object {
    "id": "1000",
    "title": "AWS",
  },
]
`;
